doc/man/pvfs2-set-debugmask.1
.TH PVFS2-SET-DEBUGMASK 1 2017-07-06
.SH NAME
\fBpvfs2-set-debugmask\fR \(en enable and disable debug strings to log
.SH SYNOPSIS
\fBpvfs2-set-debugmask\fR [\fB\-hv\fR] [\fB\-m \fImount_point\fR]
[\fB\-s \fIserver\fR] \fImask_list\fR
.SH DESCRIPTION
The server contains many log strings which are categorized.  This
utility sets which categories to log on the server.  The output will be
into the configured server log file.  This is distinct from the
.I PVFS2_DEBUGFILE
and
.I PVFS2_DEBUGMASK
environment variables, which modify the client debug mask.
.PP
The options are as follows:
.IP -h
Display synopsis.  This includes a list of all possible mask strings.
.IP -v
Print version number and exit.
.IP -m
Specify the file system to operate on.
.IP -s
Specify the address of the server.  This option is used to change the
mask only on a particular server instead of the entire cluster.
.SH ENVIRONMENT
.IP PVFS2_DEBUGFILE
If set to the path of a local file, redirect debug output to it.
.IP PVFS2_DEBUGMASK
Set the OrangeFS debug mask.  Possible masks are documented in
.BR pvfs2-set-debugmask ( 1 ) \& .
.IP PVFS2TAB_FILE
If set, the full pathname for an alternate
.IR pvfs2tab
file
.SH FILES
.I /etc/pvfs2tab
.SH EXAMPLES
Log IO errors.
.PP
.RS 6n
pvfs2-set-debugmask -m /mnt io
.RE
.SH BUGS
Please submit bug reports to devel@lists.orangefs.org
.SH SEE ALSO
.BR pvfs2-set-eventmask ( 1 ),
.BR pvfs2-set-mode ( 1 ),
.BR pvfs2tab ( 5 )


ChangeLog
-----------------------

OrangeFS Release ChangeLog
-----------------------

orangefs-2.10.0 (Git Branch v.2.10.0)
https://github.com/waltligon/orangefs.git
up to commit-id 7c28362b

New functionality:

* - general updates
	1. numerous autotools related m4 files are updated.
	2. code refactored in numerous places to eliminate
	   compiler warnings from new versions of gcc.
	3. documentation updates.

* - Server
	1. updated formatting to conform to standard in simple-stripe code.

* - Clients
	1. build liborangefs by default. liborangefs provides an orangefs
	   version of posix system calls.

        2. Windows Client:
		removed security modes
	        added version resource to exe.
	        updated compiler options to output assembler listing.
	        updated installer
	        updated Event Log message files
	        add a service description when installing service.
	        changes to service startup and shutdown
	        switched messages to manifest file
	        improved config file processing.
	        retargeted test projects to Visual Studio 2022.
	        updated pvfs2-win-cp for Visual Studio 2022.
	        updated OpenSSL
		minor changes to GetVolumeInformation.
		updated GetFileInfo and CreateFile
		Improvements to error codes
		added get-file-info utility

* - bmi
	Handle IBV_EVENT_WQ_FATAL
	Handle IBV_EVENT_WQ_FATAL
	Add reference counting changes from bmi_rdma
	Build bmi_ib OR bmi_ib_exp, not both
	Add experimental BMI InfiniBand module
	Remove return_active_nic_handle()
	Refactor RDMA data structures and some functionality
	



* - general fixes
	Fix hostname message for RDMA protocol in pvfs2-genconfig
	Fix warning about using fast networks and TCP in pvfs2-genconfig
	Fix default RDMA port in pvfs2-genconfig usage message
	Add "ib_exp" protocol to pvfs2-genconfig options
	

* - Client fixes
        1. Windows Client:
		Fixed the failure to delete files in some cases.
	        fixed CreateFile by converting kernel values to usermode.
	        error code and directory-handling fixes.
	        fixed warnings in statecomp project.
		

* - bmi fixes
	Add server shutdown fix from bmi_rdma
	Add RDMA write refcnt fix from bmi_rdma
	Fix segfault issues:
             - when shutting down restarted servers
	     - Check if already disconnected in BMI_rdma_cancel()
	Add "options" parameter to BMI_rdma_initialize()
	Fix issue causing servers to hang on shutdown
	Only increment refcnt when an RDMA write is completely done

-----------------------
Development 2.9.9
orangefs-2.9.8 (Git Branch v.2.9.x)
https://github.com/waltligon/orangefs.git
up to rev 65ab0d2

This version of OrangeFS includes many bug fixes.

	1. Disable usrint by default.
	2. Fix potential memory overwrites in various utilities.
		- Client-core
		- Showcoll
		- Mkspace
		- Xattr
		- Viewdist
		- Str-utils
	3. Eliminate compiler warnings.
		- State machines
		- Trove
	4. Improve parsing of config file.
	5. Update documentation flow.
	6. Updated readmes.
	7. Migrate documentation to hugo format.
	8. Update perf mon example.



New functionality:
	1.  Added RoCE v1 support to our IB suite.


* - Server 
	1.  Fixed problems with crdirent and rename when using
	    distributed directories.
	2.  Fixed intermittent problems with updating timestamps.

* - Clients
	1.  pvfs2-stat.c
		When the object is a symlink, we now display the string length
		of the target in the size field, making the output consistent
		with the linux kernel.  Corrected call to sys-lookup so that
		symbolic links are followed correctly when -L is specified on
		the command line.
	2.  System Interface
		1.  Fixed problems with crdirent and rename when using distributed
		    directories.
		2.  Fixed intermittent problems with updating timestamps.

* - BMI
	1.  Part of adding RoCE v1 support was to make the IB port used for
	    communication configurable.

* - Software Infrastructure

* - Testing
	1.  Generic tests, #467 and #477, of the xfstest suite were removed from our
	    testing because we do not support the functionality.

	2.  Updated our testing to use version 20170929 of LTP.  Updated
	    the list of testcases to be used.

* - Kernel Module




orangefs-2.9.7 (Subversion Branch v.2.9.x)
http://dev.orangefs.org/svn/orangefs/branches/v.2.9.x
up to rev 13115 in trunk


New functionality:
	1.  Added support for Oracle Linux 7.3 with one caveat:  Cannot use
	    OrangeFS as the backing filesystem for a user-defined loop device.
	    Support will be forth coming.

	2.  Added support for large LMDB databases using DBMaxSize config
	    parameter.  

	3.  Added scripts in the examples/profile.d directory that can be placed in 
	    /etc/profile.d.  These scripts add OrangeFS locations to
	    LIBRARY_PATH, C_INCLUDE_PATH, AND LD_LIBRARY_PATH.

	4.  Added support for Fedora 26.  Certificate security is not yet
	    compatible with OpenSSL 1.1 but work is in progress.

	5.  Official OrangeFS rpm is in the Fedora distribution.  The rpm will
	    be in either Fedora 27 or Fedora 28.  The rpm will also be in
	    their EPEL repository, allowing other RHEL distributions to access
	    it.

	6.  Added --with-lmdb to the configure script, so you can specify the
	    location of your LMDB installation.  Otherwise, LMDB delivered
	    with OrangeFS will be used.  NOTE: Either way, to use LMDB instead
	    of Berkely DB, you must specify --with-db-backend=lmdb on the
	    configure line.

	7.  pvfs2-genconfig will ask you where you want to put the server config 
	    file, with the default being the <installation directory>/etc/orangefs.conf.
	    You can still put the server config filename directly on the command line.

	8.  ACL processing is now done as specified by POSIX.


* - Server  
	* - Enhanced permission checking and job scheduling for tree
	    operations.
	* - Removed memory leaks from the getattr operation and restructured
	    keyval buffer management.
	* - Corrected problem with mgmt-create-root-dir related to having a

src/apps/admin/pvfs2-set-debugmask.c
/*
 * (C) 2001 Clemson University and The University of Chicago
 *
 * See COPYING in top-level directory.
 */

#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/time.h>
#include <time.h>
#include <stdlib.h>
#include <getopt.h>

#include "pvfs2.h"
#include "pvfs2-mgmt.h"

#ifndef PVFS2_VERSION
#define PVFS2_VERSION "Unknown"
#endif

struct options
{
    char *mnt_point;
    int mnt_point_set;
    PVFS_debug_mask debug_mask;
    int debug_mask_set;
    char *single_server;
};

static struct options *parse_args(int argc, char **argv);
static void usage(int argc, char **argv);

int main(int argc, char **argv)
{
    int ret = -1;
    PVFS_fs_id cur_fs;
    struct options *user_opts = NULL;
    char pvfs_path[PVFS_NAME_MAX] = {0};
    PVFS_credential creds;
    struct PVFS_mgmt_setparam_value param_value;

    /* look at command line arguments */
    user_opts = parse_args(argc, argv);
    if (!user_opts)
    {
        fprintf(stderr, "Error: failed to parse command line arguments.\n");
        usage(argc, argv);
        return(-1);
    }

    ret = PVFS_util_init_defaults();
    if (ret < 0)
    {
        PVFS_perror("PVFS_util_init_defaults", ret);
        return(-1);
    }

    /* translate local path into pvfs2 relative path */
    ret = PVFS_util_resolve(user_opts->mnt_point,
                            &cur_fs, pvfs_path, PVFS_NAME_MAX);
    if (ret < 0)
    {
        fprintf(stderr, "Error: could not find filesystem "
                "for %s in pvfstab\n",  user_opts->mnt_point);
        return(-1);
    }

    ret = PVFS_util_gen_credential_defaults(&creds);
    if (ret < 0)
    {
        PVFS_perror("PVFS_util_gen_credential_defaults", ret);
        return(-1);
    }

    if (user_opts->single_server)
    {
        printf("Setting debugmask on server %s\n",
               user_opts->single_server);

        param_value.type = PVFS_MGMT_PARAM_TYPE_MASK;
        param_value.u.mask_value = user_opts->debug_mask;
        ret = PVFS_mgmt_setparam_single(cur_fs,
                                        &creds,
                                        PVFS_SERV_PARAM_GOSSIP_MASK,
                                        &param_value,
                                        user_opts->single_server,
                                        NULL,
                                        NULL);
    }
    else
    {
        printf("Setting debugmask on all servers\n");

        param_value.type = PVFS_MGMT_PARAM_TYPE_MASK;
        param_value.u.mask_value = user_opts->debug_mask;
        ret = PVFS_mgmt_setparam_all(cur_fs,
                                     &creds,
                                     PVFS_SERV_PARAM_GOSSIP_MASK,
                                     &param_value,
                                     NULL,
                                     NULL);
    }

    if (ret)
    {
        char buf[64] = {0};
        PVFS_strerror_r(ret, buf, 64);
        fprintf(stderr, "Setparam failure: %s (%d)\n", buf, ret);
    }
    return PVFS_sys_finalize();
}

/* parse_args()
 *
 * parses command line arguments
 *
 * returns pointer to options structure on success, NULL on failure
 */
static struct options *parse_args(int argc, char **argv)
{
    int ret = -1, len = 0, option_index = 0;
    const char *cur_option = NULL;
    struct options *tmp_opts = NULL;
    static struct option long_opts[] =
    {
        {"help",0,0,0},
        {"version",0,0,0},
        {"mount",1,0,0},
        {"server",1,0,0},
        {0,0,0,0}
    };

    /* create storage for the command line options */
    tmp_opts = (struct options*)malloc(sizeof(struct options));
    if (!tmp_opts)
    {
        return(NULL);
    }
    memset(tmp_opts, 0, sizeof(struct options));

    /* fill in defaults (except for hostid) */
    tmp_opts->debug_mask = GOSSIP_NO_DEBUG;

    while((ret = getopt_long(argc, argv, "hvm:s:",
                             long_opts, &option_index)) != -1)
    {
        switch(ret)
        {
            case 0:
                cur_option = long_opts[option_index].name;

                if (strcmp("help", cur_option) == 0)
                {
                    goto do_help;
                }
                else if (strcmp("version", cur_option) == 0)
                {
                    goto do_version;
                }
                else if (strcmp("mount", cur_option) == 0)
                {
                    goto do_mount;
                }
                else if (strcmp("server", cur_option) == 0)
                {
                    goto do_server;
                }
                else
                {
                    usage(argc, argv);
                    exit(-1);
                }
                break;
            case('h'):
          do_help:
                usage(argc, argv);
                exit(0);
            case('v'):
          do_version:
                printf("%s\n", PVFS2_VERSION);
                exit(0);
            case('s'):
          do_server:
                tmp_opts->single_server = strdup(optarg);
                if (!tmp_opts)
                {
                    return NULL;
                }
                break;
            case('m'):
          do_mount:
                len = strlen(optarg) + 1;
                tmp_opts->mnt_point = (char*)malloc(len+1);
                if (!tmp_opts->mnt_point)
                {
                    free(tmp_opts);
                    return NULL;
                }
                memset(tmp_opts->mnt_point, 0, len+1);
                ret = sscanf(optarg, "%s", tmp_opts->mnt_point);
                if(ret < 1)
                {
                    free(tmp_opts);
                    return NULL;
                }
                /* TODO: dirty hack... fix later.  The remove_dir_prefix()
                 * function expects some trailing segments or at least
                 * a slash off of the mount point
                 */
                strcat(tmp_opts->mnt_point, "/");
                tmp_opts->mnt_point_set = 1;
                break;
            case('?'):
                usage(argc, argv);
                exit(-1);
        }
    }

    if (optind != (argc - 1))
    {
        usage(argc, argv);
        exit(-1);
    }

    tmp_opts->debug_mask = PVFS_debug_eventlog_to_mask(argv[argc-1]);
    tmp_opts->debug_mask_set = 1;

    if (!tmp_opts->mnt_point_set)
    {
        free(tmp_opts);
        tmp_opts = NULL;
    }
    return tmp_opts;
}


static void usage(int argc, char **argv)
{
    int i = 0;
    int j = 0;
    const char *mask = NULL;
    int max_dbg_keyword_len = 0;
    int spacing = 0;

    fprintf(stderr, "Usage: %s [OPTION] <mask_list>\n\n",
            argv[0]);
    fprintf(stderr, "-h, --help                  display this help "
            "information and exit\n");
    fprintf(stderr, "-v, --version               display version "
            "information and exit\n");
    fprintf(stderr, "-m, --mount=<mount_point>   use the specified "
            "mount point\n");
    fprintf(stderr, "-s, --server=<server_addr>  use the specified "
            "server address\n");
    fprintf(stderr, "\nExamples:\n  %s -m /mnt/pvfs2 \"network,server"
            "\"\n", argv[0]);
    fprintf(stderr, "\t[ enables network and server debugging on all "
            "servers ]\n\n");
    fprintf(stderr, "  %s -m /mnt/pvfs2 \"none\" -s \"tcp://localhost:"
            "3334/pvfs2-fs\"\n", argv[0]);
    fprintf(stderr, "\t[ disable debugging on the specified server only "
            " ]\n\n");
    fprintf(stderr, "Available masks include:\n");
    
    while((mask = PVFS_debug_get_next_debug_keyword(i++)) != NULL)
    {
        int length = strlen(mask);
        if(length > max_dbg_keyword_len)
        {
            max_dbg_keyword_len = length;
        }
    }

    i = 0;
    mask = NULL;
    spacing = max_dbg_keyword_len + 4;
    while((mask = PVFS_debug_get_next_debug_keyword(i++)) != NULL)
    {
        if((i % 4) == 1)
        {
            fprintf(stderr, "    ");
        }
        fprintf(stderr,"%s",mask);
        if((i % 4) != 0)
        {
            for(j = 0; j < (spacing - strlen(mask)); j++)
            {
                putc((int) ' ', stderr);
            }
        }
        if ((i % 4) == 0)
        {
            fprintf(stderr,"\n");
        }
    }
    fprintf(stderr,"\n");
}

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
